### PR TITLE
Add Support for Layer Titles

### PR DESCRIPTION
**Pull request by theduckylittle**
1. Moved _sourcesVersion, _layersVersion, and _dataVersion
   into the metadata object.  I used "bnd:" as the Boundless prefix
   to prevent clashing. (MapBox recommends prefixes in the style spec.)
2. Cleaned up some linting inconsistencies "if/for(" instead of "if/for (".
3.  Added support for layer titles.
   This support will be necessary come time for making a simple layer list.

### CODE REVIEW

File: __tests__/components/map.test.js [hunk 1]
```diff
@@ -44,10 +44,14 @@ describe('Map component', () => {
       "id": "states",
       "source": "states-wms",
     }];
+    const metadata = {
+      'bnd:sources-verison': 0,
```
Comment (bartvde) [inline __tests__/components/map.test.js]: typo: verison instead of version

File: __tests__/components/map.test.js [hunk 2]
```diff
@@ -81,8 +85,11 @@ describe('Map component', () => {
     }];
     const center = [0, 0];
     const zoom = 2;
-    const _sourcesVersion = 0, _layersVersion = 0;
-    const wrapper = shallow(<Map map={{center, zoom, sources, layers, _sourcesVersion, _layersVersion}} />);
+    const metadata = {
+      'bnd:sources-verison': 0,
```
Comment (bartvde) [inline __tests__/components/map.test.js]: typo

File: __tests__/components/map.test.js [hunk 3]
```diff
@@ -103,10 +110,14 @@ describe('Map component', () => {
       "id": "tilejson-layer",
       "source": "tilejson"
     }];
+
+    const metadata = {
+      'bnd:sources-verison': 0,
```
Comment (bartvde) [inline __tests__/components/map.test.js]: typo

File: __tests__/components/map.test.js [hunk 4]
```diff
@@ -126,10 +137,14 @@ describe('Map component', () => {
...
```
Comment (bartvde) [inline __tests__/components/map.test.js]: typo

File: __tests__/components/map.test.js [hunk 5]
```diff
@@ -139,8 +154,10 @@ describe('Map component', () => {
       map: {
         center,
         zoom,
-        _sourcesVersion: 0,
-        _layersVersion: 1,
+        metadata: {
+          'bnd:sources-verison': 0,
```
Comment (bartvde) [inline __tests__/components/map.test.js]: typo

File: __tests__/components/map.test.js [hunk 6]
```diff
@@ -168,8 +185,11 @@ describe('Map component', () => {
     }];
     const center = [0, 0];
     const zoom = 2;
-    const _sourcesVersion = 0, _layersVersion = 0;
-    const wrapper = shallow(<Map map={{center, zoom, sources, layers, _sourcesVersion, _layersVersion}} />);
+    const metadata = {
+      'bnd:sources-verison': 0,
```
Comment (bartvde) [inline __tests__/components/map.test.js]: typo

File: __tests__/components/map.test.js [hunk 7]
```diff
@@ -178,8 +198,10 @@ describe('Map component', () => {
       map: {
         center,
         zoom,
-        _sourcesVersion: 0,
-        _layersVersion: 1,
+        metadata: {
+          'bnd:sources-verison': 0,
```
Comment (bartvde) [inline __tests__/components/map.test.js]: typo

File: __tests__/components/map.test.js [hunk 8]
```diff
@@ -190,8 +212,10 @@ describe('Map component', () => {
       map: {
         center,
         zoom,
-        _sourcesVersion: 0,
-        _layersVersion: 2,
+        metadata: {
+          'bnd:sources-verison': 0,
```
Comment (bartvde) [inline __tests__/components/map.test.js]: typo

File: __tests__/reducers/map.test.js [hunk 9]
```diff
@@ -259,16 +340,18 @@ describe('map reducer', () => {
...
+        'bnd:layers-version': 0,
+      },
       layers: []
     })
   })
 
   it('should handle ADD_FEATURES', () => {
     // since we do not go through ADD_SOURCE we need to set _dataVersion
-    const source = {"_dataVersion": 0, "data": {"type":"FeatureCollection","crs":{"type":"name","properties":{"name":"urn:ogc:def:crs:OGC:1.3:CRS84"}},"features":[{"type":"Feature","properties":{"n":2,"cat":1},"geometry":{"type":"Point","coordinates":[0.5,0.5]}},{"type":"Feature","properties":{"n":3,"cat":2},"geometry":{"type":"Point","coordinates":[0.5,1.5]}}]}};
+    const source = {"metadata" : {"bnd:data-version" : 0}, "data": {"type":"FeatureCollection","crs":{"type":"name","properties":{"name":"urn:ogc:def:crs:OGC:1.3:CRS84"}},"features":[{"type":"Feature","properties":{"n":2,"cat":1},"geometry":{"type":"Point","coordinates":[0.5,0.5]}},{"type":"Feature","properties":{"n":3,"cat":2},"geometry":{"type":"Point","coordinates":[0.5,1.5]}}]}};
```
Comment (bartvde) [inline __tests__/reducers/map.test.js]: it looks like mapbox only defines metadata at the layer and the root level officially?

Comment (bartvde) [inline __tests__/reducers/map.test.js]: So maybe we should store it somewhere in the root metadata keyed by the source id?

File: src/actions/map.js [hunk 10]
```diff
@@ -12,10 +12,10 @@ export function setView(center, zoom) {
   }
 }
 
-export function addLayer(layerDef) {
+export function addLayer(layerDef, layerTitle) {
   return {
     type: MAP.ADD_LAYER,
-    layerDef
+    layerDef, layerTitle
```
Comment (bartvde) [inline src/actions/map.js]: we could also just provide it in the layerDef directly in the metadata or not? Wonder why it was separated, could be a good reason for it ofcourse, but I couldn't think of one

File: src/reducers/map.js [hunk 11]
```diff
@@ -23,71 +39,90 @@ function addLayer(state, action) {
...
+  // add a layer name if specified in the action.
+  if (action.layerTitle) {
+    new_layer.metadata['bnd:title'] = action.layerTitle;
+  }
+
   return Object.assign({}, state, {
-    _layersVersion: state._layersVersion + 1,
     layers: state.layers.concat([new_layer]),
-  });
+  }, incrementVersion(state.metadata, 'bnd:layers-version'));
```
Comment (bartvde) [inline src/reducers/map.js]: maybe we should put this a const on the top of the file?

```const layerVersionKey = 'bnd:layers-version'``` and use that const everywhere else in the code?

File: src/actions/map.js (hunk 10 above)
Comment (theduckylittle) [inline src/actions/map.js]: Honestly, I did it as a convenience wrapper for the developer.  The reducer handles converting it to the metadata syntax.  I just through having a layer title would be a common enough situation that easing its use was worthwhile.

Comment (bartvde) [inline src/actions/map.js]: I'm fine with keeping it this way

**Comment by theduckylittle** [comment]
@bartvde I've updated the code to use constants instead of dropping strings everywhere.  I made some decisions that I would love some feedback about:

1. I created a function called "dataVersionKey" that both the reducer and the map component use.  I  imported the function FROM the reducer into the map module.  I sorta feel like that shouldn't happen and maybe we should cave into creating a "util.js" or similar module for functions that are useful in multiple places.
2. The 'constants.js' file is another one of those 'catch-all' modules.  I did that BEFORE writing the dataVersionKey function.  Maybe 'constants.js' could be moved to a 'constants' object that is apart of a 'util.js' module? Maybe a 'metadata.js' module would be appropriate as both the function and the constants strictly concern metadata-keys.

Concerning the linting: I will get a proper eslint configuration together -- I have been linting manually thus far.

**Comment by bartvde** [comment]
Yeah I think an util module makes sense. I don't have a strong opinion on if constants should be a separate file, I think it is fine as you've done it now.

**Pull request merged**